feat(ivy): use i18n locale data to determine the plural form of ICU expressions

ocombe opened the pull request:
Plural ICU expressions depend on the locale (different languages have different plural forms). Until now the locale was hard coded as `en-US`.
For compatibility reasons, if you use ivy with AOT and bootstrap your app with `bootstrapModule` then the `LOCALE_ID` token will be set automatically for ivy, which is then used to get the correct plural form.
If you use JIT, you need to define the `LOCALE_ID` provider on the module that you bootstrap.
For `TestBed` you can use either `configureTestingModule` or `overrideProvider` to define that provider.
If you don't use the compat mode and start your app with `renderComponent` you need to call `ɵsetLocaleId` manually to define the `LOCALE_ID` before bootstrap. We expect this to change once we start adding the new i18n APIs, so don't rely on this function (there's a reason why it's a private export).

IgorMinar reviewed (commented):
please reexport the symbols and types and reattach the docs to the exports.

afterwards please validate that the api docs look right (the type signatures are correct) on the ngbuilds.io preview server.

kara commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -2104,4 +2107,56 @@ describe('Runtime i18n', () => {
     });
   });
 
+  it('should return the correct plural form for ICU expressions when using a specific locale',
```
Can you re-write this as a TestBed test and move to the i18n_integration_spec? We shouldn't have any new tests with handwritten generated code.

ocombe replied:
Well actually the locale data def (`ngLocaleIdDef`) isn't defined for test bed, so that wouldn't work :-/

ocombe replied:
I've added a TestBed test, but I kept this one as they don't test the same thing

mary-poppins commented:
You can preview d9ff164 at https://pr29249-d9ff164.ngbuilds.io/.
You can preview edc9394 at https://pr29249-edc9394.ngbuilds.io/.
You can preview a9a43c8 at https://pr29249-a9a43c8.ngbuilds.io/.
You can preview d204e5d at https://pr29249-d204e5d.ngbuilds.io/.

mary-poppins commented:
You can preview 6b3c5da at https://pr29249-6b3c5da.ngbuilds.io/.

ocombe commented:
So I updated the code as requested by @IgorMinar so that it re-exports the symbols. It works for api guard, but we loose the api docs in the process.
![image](https://user-images.githubusercontent.com/265378/55011135-c2740800-4fe5-11e9-8086-550ff0759393.png)

@petebacondarwin seems to think that it will be complicated to fix dgeni for that... not sure what the next move is.
We can duplicate the api docs, but we will still see the wrong code: `type FormStyle = ɵFormStyle;` instead of
```enum FormStyle {
  Format
  Standalone
}```

ocombe commented:
Blocked by FW-1207

mary-poppins commented:
You can preview bf02a00 at https://pr29249-bf02a00.ngbuilds.io/.

mary-poppins commented:
You can preview 66f74e7 at https://pr29249-66f74e7.ngbuilds.io/.

alxhub commented on packages/core/test/application_ref_spec.ts:
```diff
@@ -326,6 +327,28 @@ class SomeComponent {
         expect(loadResourceSpy).toHaveBeenCalledWith('/test-template.html');
       });
+
+      onlyInIvy('We only need to define `LOCALE_ID` for runtime i18n')
```
Do we need `ResourceLoader` as part of this test? It could be written without it.

ocombe replied:
yes because `compileNgModuleFactory` requires `ResourceLoader`, which is usually provided when you use platform-browser in a real application, but not here in the tests. It throws with an error otherwise...

alxhub replied:
I think it's only needed here because the test was copied from another which uses a `templateUrl`. If you don't use a `templateUrl` with the component you shouldn't need the `ResourceLoader`.

alxhub commented on packages/core/test/i18n_integration_spec.ts:
```diff
@@ -644,7 +643,7 @@ onlyInIvy('Ivy i18n logic').describe('i18n', function() {
      () => {
        registerLocaleData(localeRo);
-       setLocaleId('ro');
+       TestBed.overrideProvider(LOCALE_ID, {useValue: 'ro'});
```
```suggestion
       TestBed.configureTestingModule({providers: [LOCALE_ID, {useValue: 'ro'}]});
```

ocombe replied:
We already use `configureTestingModule` in `beforeEach`, if I use it here I'll have to re-add the declarations

alxhub replied:
`configureTestingModule` is cumulative.

alxhub commented on packages/core/testing/src/r3_test_bed_compiler.ts:
```diff
@@ -41,10 +41,12 @@ import {
   ɵDirectiveDef as DirectiveDef,
   ɵpatchComponentDefWithScope as patchComponentDefWithScope,
-  ɵtransitiveScopesFor as transitiveScopesFor,
+  ɵtransitiveScopesFor as transitiveScopesFor, LOCALE_ID,
```
Sort this into the import list on its own line, please :)

alxhub commented on packages/core/testing/src/r3_test_bed_compiler.ts:
```diff
@@ -41,10 +41,12 @@ import {
 } from '@angular/core';
 // clang-format on
 import {ResourceLoader} from '@angular/compiler';
+import {setLocaleId} from '@angular/core/src/render3';
```
We should try to avoid deep imports. Can these be exported (publicly or privately) from the root of `@angular/core`?

ocombe replied:
it is privately exported, updated the code to use that

alxhub commented on packages/core/testing/src/r3_test_bed_compiler.ts:
```diff
@@ -264,6 +266,9 @@ export class R3TestBedCompiler {
     this.testModuleRef = new NgModuleRef(this.testModuleType, parentInjector);
 
+    // Set the locale ID, it can be overridden for the tests
+    const localeId = this.testModuleRef.injector.get(LOCALE_ID);
```
Should we specify a default here, like we do in `compileNgModuleFactory`?

alxhub reviewed (changes requested):
I think the logic in `compileNgModuleFactory` might be wrong.

alxhub commented on packages/core/src/application_ref.ts:
```diff
@@ -261,6 +263,9 @@ export class PlatformRef {
       if (!exceptionHandler) {
         throw new Error('No ErrorHandler. Is platform module (BrowserModule) included?');
       }
+      // If the `LOCALE_ID` provider is defined at bootstrap, we set the value for runtime i18n
```
Maybe expand the comment to mention that this is an Ivy-specific operation?

mary-poppins commented:
You can preview fa18cec at https://pr29249-fa18cec.ngbuilds.io/.

mary-poppins commented:
You can preview 36de339 at https://pr29249-36de339.ngbuilds.io/.

mary-poppins commented:
You can preview 40445bb at https://pr29249-40445bb.ngbuilds.io/.

mhevery commented:
http://test/OCL:242135545:BASE:242135584:1554480750448:5f91d5e9

mhevery commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -6,685 +6,181 @@
-  Standalone
-}
-
+export type FormatWidth = ɵFormatWidth;
```
The fact that you have to use `FormatWidth_` is annoying. So I think if you move it into a separate file it would work just fine.

ocombe replied:
This doesn't work, typescript errors with `Argument of type 'FormatWidth_' is not assignable with parameter of type 'FormatWidth'` whenever I tried to use one of the functions exported by core (and re-exported by common), for example if I do `const formatValue = getLocaleDateFormat(locale, FormatWidth.Short);`

mhevery reviewed (changes requested):
Public API guard needs to be fixed.

mary-poppins commented:
You can preview 670ecc7 at https://pr29249-670ecc7.ngbuilds.io/.

mary-poppins commented:
You can preview bcd6fdf at https://pr29249-bcd6fdf.ngbuilds.io/.

mhevery reviewed (changes requested):
Your change will make it so that `core:Plural.ZERO !== common:Plural.ZERO` which is wrong.

61a4dd9304 is a proper fix. Please align rest of the types with it.

mary-poppins commented:
You can preview 61a4dd9 at https://pr29249-61a4dd9.ngbuilds.io/.

mary-poppins commented:
You can preview 956617f at https://pr29249-956617f.ngbuilds.io/.

mhevery commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -495,8 +496,7 @@ function getLocaleCurrencies(locale: string): {[code: string]: CurrenciesSymbols
  * @publicApi
  */
 export function getLocalePluralCase(locale: string): (value: number) => Plural {
```
We should not have to have wrapping functions. Can't we just do:
```
export const getLocalePluralCase: .... = ɵgetLocalePluralCase;
```

ocombe replied:
if we do that, we loose the public api guard definition (it'll just be an alias instead of the full definition), which would relax our constraints

ocombe replied:
actually let me try something

AndrewKushnir reviewed (approved):
LGTM 👍

mhevery commented on packages/core/src/i18n/localization.ts:
```diff
@@ -15,15 +15,15 @@ export function getPluralCase(value: any, locale: string): string {
 
   switch (plural) {
-    case Plural.Zero:
+    case 0:
```
I think you can revert this. The original code was better.

mhevery commented on packages/common/src/common.ts:
```diff
@@ -15,8 +15,8 @@ export * from './location/index';
 export {NgLocaleLocalization, NgLocalization} from './i18n/localization';
-export {registerLocaleData, Plural} from './i18n/locale_data';
-export {NumberFormatStyle, FormStyle, Time, TranslationWidth, FormatWidth, NumberSymbol, WeekDay, getNumberOfCurrencyDigits, getCurrencySymbol, getLocaleDayPeriods, getLocaleDayNames, getLocaleMonthNames, getLocaleId, getLocaleEraNames, getLocaleWeekEndRange, getLocaleFirstDayOfWeek, getLocaleDateFormat, getLocaleDateTimeFormat, getLocaleExtraDayPeriodRules, getLocaleExtraDayPeriods, getLocalePlu ...
+export {registerLocaleData} from './i18n/locale_data';
```
The issue here is that `Plural` defined in `locale_data` is type information only, no value. So when rollup tries to read the value it blows up.

mhevery commented on packages/common/src/common.ts:
```diff
@@ -15,8 +15,8 @@ export * from './location/index';
-export {registerLocaleData, Plural} from './i18n/locale_data';
-export {NumberFormatStyle, FormStyle, Time, TranslationWidth, FormatWidth, NumberSymbol, WeekDay, getNumberOfCurrencyDigits, getCurrencySymbol, getLocaleDayPeriods, getLocaleDayNames, getLocaleMonthNames, getLocaleId, getLocaleEraNames, getLocaleWeekEndRange, getLocaleFirstDayOfWeek, getLocaleDateFormat, getLocaleDateTimeFormat, getLocaleExtraDayPeriodRules, getLocaleExtraDayPeriods, getLocalePlu ...
+export {registerLocaleData} from './i18n/locale_data';
+export {Plural, NumberFormatStyle, FormStyle, Time, TranslationWidth, FormatWidth, NumberSymbol, WeekDay, getNumberOfCurrencyDigits, getCurrencySymbol, getLocaleDayPeriods, getLocaleDayNames, getLocaleMonthNames, getLocaleId, getLocaleEraNames, getLocaleWeekEndRange, getLocaleFirstDayOfWeek, getLocaleDateFormat, getLocaleDateTimeFormat, getLocaleExtraDayPeriodRules, getLocaleExtraDayPeriods, getL ...
```
`Plural` here contains both type as well as value, hence rollup works.

kara commented:
[presubmit](http://test/OCL:242713752:BASE:242713809:1554836806793:8832bbc)

IgorMinar commented:
the last presubmit failed - running an new one: http://test/OCL:243070544:BASE:243070563:1554995271004:b929419d

IgorMinar commented:
@ocombe can you please rebase?

ocombe commented:
@IgorMinar done

alxhub commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:243629037:BASE:243630442:1555345847096:3f2f1f27) after rebase.

Caretaker note: this PR requires an additional g3 fix to its BUILD file.

alxhub commented:
Caretakers: this PR was reverted originally due to its effect on a number of Firebase targets in g3 (it caused build timeouts). This needs to be investigated.

alxhub reviewed (changes requested):
This PR is being held until the start of partner testing. Talk to @alxhub or @kara for context.

ocombe committed:
feat(ivy): use i18n locale data to determine the plural form of ICU expressions

Plural ICU expressions depend on the locale (different languages have different plural forms). Until now the locale was hard coded as `en-US`.
For compatibility reasons, if you use ivy with AOT and bootstrap your app with `bootstrapModule` then the `LOCALE_ID` token will be set automatically for ivy, which is then used to get the correct plural form.
If you use JIT, you need to define the `LOCALE_ID` provider on the module that you bootstrap.
For `TestBed` you can use either `configureTestingModule` or `overrideProvider` to define that provider.
If you don't use the compat mode and start your app with `renderComponent` you need to call `ɵsetLocaleId` manually to define the `LOCALE_ID` before bootstrap. We expect this to change once we start adding the new i18n APIs, so don't rely on this function (there's a reason why it's a private export).

ocombe commented:
The size increase check is false (I manually checked against master myself, and the bundle file for hello_world is exactly the same)

jasonaden commented:
[Presubmit](http://test/OCL:249071179:BASE:249071208:1558371940116:484e87c6)

matsko commented:
@ocombe rebase please

matsko commented:
http://test/OCL:250167572:BASE:250167579:1558978723403:5c4448db
